packages/multichain/src/adapters/caip-permission-adapter-eth-accounts.ts

```diff
@@ -0,0 +1,125 @@
+  const { namespace } = parseScopeString(scopeString);
+
+  return (
+    namespace === KnownCaipNamespace.Eip155 ||
+    scopeString === KnownWalletScopeString.Eip155
+  );
+};
+
+/**
+ * Gets the Ethereum (EIP155 namespaced) accounts from the required and optional scopes.
+ * @param caip25CaveatValue - The CAIP-25 caveat value to get the Ethereum accounts from.
+ * @returns An array of Ethereum accounts.
+ */
+export const getEthAccounts = (
+  caip25CaveatValue: Pick<
+    Caip25CaveatValue,
+    'requiredScopes' | 'optionalScopes'
+  >,
+) => {
+  const ethAccounts: string[] = [];
+  const sessionScopes = mergeScopes(
+    caip25CaveatValue.requiredScopes,
+    caip25CaveatValue.optionalScopes,
+  );
+
+  Object.entries(sessionScopes).forEach(([_, { accounts }]) => {
+    accounts?.forEach((account) => {
+      const { address, chainId } = parseCaipAccountId(account);
+
+      if (isEip155ScopeString(chainId)) {
+        ethAccounts.push(address);
+      }
+    });
+  });
+
+  return getUniqueArrayItems(ethAccounts);
+};
+
+/**
+ * Sets the Ethereum (EIP155 namespaced) accounts for the given scopes object.
+ * @param scopesObject - The scopes object to set the Ethereum accounts for.
+ * @param accounts - The Ethereum accounts to set.
+ * @returns The updated scopes object with the Ethereum accounts set.
+ */
+const setEthAccountsForScopesObject = (
+  scopesObject: InternalScopesObject,
+  accounts: Hex[],
+) => {
+  const updatedScopesObject: InternalScopesObject = {};
+
+  Object.entries(scopesObject).forEach(([scopeString, scopeObject]) => {
+    const isWalletNamespace = scopeString === KnownCaipNamespace.Wallet;
+
+    if (
+      !isEip155ScopeString(scopeString as InternalScopeString) &&
+      !isWalletNamespace
+    ) {
+      updatedScopesObject[scopeString as InternalScopeString] = scopeObject;
+      return;
+    }
+
+    const caipAccounts = accounts.map(
+      (account) =>
+        (isWalletNamespace
+          ? `${KnownWalletScopeString.Eip155}:${account}`
+          : `${scopeString}:${account}`) as CaipAccountId,
+    );
+
+    updatedScopesObject[scopeString as InternalScopeString] = {
+      ...scopeObject,
+      accounts: caipAccounts,
+    };
+  });
+
+  return updatedScopesObject;
+};
+
+/**
+ * Sets the Ethereum (EIP155 namespaced) accounts for the given CAIP-25 caveat value.
+ * @param caip25CaveatValue - The CAIP-25 caveat value to set the Ethereum accounts for.
+ * @param accounts - The Ethereum accounts to set.
+ * @returns The updated CAIP-25 caveat value with the Ethereum accounts set.
+ */
+export const setEthAccounts = (
+  caip25CaveatValue: Caip25CaveatValue,
+  accounts: Hex[],
+) => {
+  return {
+    ...caip25CaveatValue,
+    requiredScopes: setEthAccountsForScopesObject(
+      caip25CaveatValue.requiredScopes,
+      accounts,
+    ),
+    optionalScopes: setEthAccountsForScopesObject(
+      {
+        [KnownWalletScopeString.Eip155]: {
```

Comment:
Sorry I'm not following what's happening here.

This method is setting accounts on any EIP-155 scopes present, required or optional. But we're making a point here to add an optional scope if it doesn't exist already.

I am guessing that we're applying the accounts equally across all scopes because we intentionally are not supporting "specific accounts for specific chains" in our permission flow for the moment. That makes sense (though this is worth clarifying in the TSDoc description)

But why add an optional EIP-155 scope when it doesn't exist already?

Reply:
Its a good question. I'm trying to remember what the thinking was on this 🤔

Reply:
Ok right so I think this is also related to our short term need to provide a path for account permissions without chain permissions for snaps. Until we introduce chain permissions to snaps we need to park them in the `KnownWalletScopeString.Eip155`

Reply:
I see, OK. Makes sense. So to summarize:
* Synchronizing accounts across all EIP-155 scopes, because our UI doesn't support per-chain scopes, and
* Creating this wallet EIP-155 scope to support snaps

Could we document this somewhere, perhaps in the TSDoc comment?

Reply:
updated the TSDoc per your suggestion as well: https://github.com/MetaMask/core/pull/4784/commits/5371ff0d09fa463e841d5cb2f0e8db6d83400290

Reply:
added to the doc to explain why we add the `wallet:eip155` scope here: https://github.com/MetaMask/core/pull/4784/commits/bce936014cf00c4ec3019898a42a581c4f93247b